kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/impl/cache/CacheImpl.java

```diff
@@ -111,10 +97,27 @@ public synchronized T put(T obj) {
     }
     String key = getKey(obj);
     T oldObj = this.items.put(key, obj);
+    int comp = compareResourceVersions(oldObj, obj);
+    if (comp == 0) {
+      return obj;
+    }
+    if (comp > 0) {
+      // put the old one back, it's later
+      this.items.put(key, oldObj);
+      return obj;
+    }
     this.updateIndices(oldObj, obj, key);
     return oldObj;
   }
 
+  static <T extends HasMetadata> int compareResourceVersions(T oldObj, T obj) {
+    if (oldObj == null) {
+      return -1;
+    }
+    return Long.compare(Long.parseLong(oldObj.getMetadata().getResourceVersion()),
```

Comment:
Even if I have never seen anything other than increasing numbers as resource versions, I think this is making too much assumption regarding the specification.
As declared in the official API, the resourceVersion is `an opaque value that represents the internal version of this object`, cf https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#objectmeta-v1-meta
As far as I can tell, the only thing we can do to tell if two versions are different, not order them.

Reply:
I agree that it's not inline with the official docs, but this is how resourceVersion will function until etcd is no longer being utilized.  You can see similar logic in the go client https://github.com/kubernetes/client-go/blame/45fb834fedea5c753501ea8ea597e2dc54080510/tools/cache/mutation_cache.go#L242

Obviously for your usage, you don't need this checking - if modifications are restricted to before the informer is running there's no consistency issue.  However I've been asked by the operator sdk team about also supporting proactive updates to the cache as well - and I really can't come up with any other to do that consistently.  My hope was to just address both of those cases together and providing a writable store in general.

The options are:
1. Offer an option to, or just allow, unsafe updates - I don't believe that's a good approach though
2. Restrict this checking to only when someone has made mutations after calling run and document that it may not always be supported
3. Only support modifications before run and still don't try to support consistent modifications after run has been called.

One more thing that I would need to add is that if this checking is done, is that relist needs to force updates.  In cases like this https://github.com/kubernetes/kubernetes/issues/107457 the restart of the api server will cause a relist and we'll see older resourceVersions - this is obviously an edge case which currently does't work even with the go client informers that are relisting with a resourceVersion.

Reply:
I'll update this to number 2 and also correct the update/delete to be atomic and then we can see what everyone thinks of that.

Reply:
Ok, I didn't know that resource versions were number due to etcd, nice to know. I am just a user of Kubernetes, so I can only relay, but I have a heard that for scaling kubernetes some companies have setup other kind of storage like postgres. Considering the code in the go client you pointed, I guess that even with other storage they ensured that resources version are still incresing numbers.
My biggest worry was about `Long.parseLong` raising exception and crashing all that.
Thank you for sharing.

Reply:
Looking at the additional complexity of accounting for forcing updates on relist and cases like https://github.com/kubernetes/kubernetes/issues/107457 - I'm having second thoughts as it's still not possible to ensure that your consistent with the api server.  So I'll have to keep maintaining it's not a good idea to manipulate the store on a running informer.  I'll refine this pr to 3 instead.

Reply:
Linking to #https://github.com/fabric8io/kubernetes-client/pull/3943#issuecomment-1067962930 and resolving this comment.  If the operator sdk want to pursue this logic, it will need to do so with an adjunct cache.